vnext/Microsoft.ReactNative/JsiApi.idl

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.ReactNative
+{
+  interface IJsiRuntime;
+
+  delegate void JsiDataHandler(UInt8[] data);
+
+  [webhosthidden]
+  interface IJsiBuffer
+  {
+    UInt32 Size { get; };
+    void GetData(JsiDataHandler handler);
+  };
+
+  enum JsiValueKind
+  {
+    Undefined,
+    Null,
+    Boolean,
+    Number,
+    Symbol,
+    String,
+    Object,
+  };
+
+  [webhosthidden]
+  struct JsiValueData
+  {
+    JsiValueKind Kind;
+    UInt64 Data;
+  };
+
+  struct JsiSymbolData
+  {
+    UInt64 Data;
+  };
+
+  struct JsiStringData
+  {
+    UInt64 Data;
+  };
+
+  struct JsiObjectData
+  {
+    UInt64 Data;
+  };
+
+  struct JsiWeakObjectData
+  {
+    UInt64 Data;
+  };
+
+  struct JsiPropertyNameIdData
+  {
+    UInt64 Data;
+  };
+
+  struct JsiFunctionData
+  {
+    UInt64 Data;
+  };
+
+  struct JsiArrayData
+  {
+    UInt64 Data;
+  };
+
+  struct JsiArrayBufferData
+  {
+    UInt64 Data;
+  };
+
+  struct JsiScopeState
+  {
+    UInt64 Data;
+  };
+
+  [default_interface]
+  runtimeclass JsiPreparedJavaScript
+  {
+  };
+
+  delegate JsiValueData JsiHostFunction(IJsiRuntime runtime, JsiValueData thisArg, JsiValueData[] args);
+
+  [webhosthidden]
+  interface IJsiHostObject
+  {
+    JsiValueData GetProperty(IJsiRuntime runtime, JsiPropertyNameIdData name);
+    void SetProperty(IJsiRuntime runtime, JsiPropertyNameIdData name, JsiValueData value);
+    IVector<JsiPropertyNameIdData> GetPropertyNames(IJsiRuntime runtime);
+  };
+
+  [webhosthidden]
+  interface IJsiRuntime
+  {
+    JsiValueData EvaluateJavaScript(IJsiBuffer buffer, String sourceUrl);
+    JsiPreparedJavaScript PrepareJavaScript(IJsiBuffer buffer, String sourceUrl);
+    JsiValueData EvaluatePreparedJavaScript(JsiPreparedJavaScript js);
+    JsiObjectData Global { get; };
+    String Description { get; };
+    Boolean IsInspectable { get; };
+
+    JsiSymbolData CloneSymbol(JsiSymbolData symbol);
+    JsiStringData CloneString(JsiStringData str);
+    JsiObjectData CloneObject(JsiObjectData obj);
+    JsiPropertyNameIdData ClonePropertyNameId(JsiPropertyNameIdData propertyNameId);
+
+    JsiPropertyNameIdData CreatePropertyNameIdFromAscii(UInt8[] ascii);
+    JsiPropertyNameIdData CreatePropertyNameIdFromUtf8(UInt8[] utf8);
+    JsiPropertyNameIdData CreatePropertyNameIdFromString(JsiStringData str);
+    void PropertyNameIdToUtf8(JsiPropertyNameIdData propertyNameId, JsiDataHandler utf8Handler);
+    Boolean PropertyNameIdEquals(JsiPropertyNameIdData left, JsiPropertyNameIdData right);
+
+    void SymbolToUtf8(JsiSymbolData symbol, JsiDataHandler utf8Handler);
+
+    JsiStringData CreateStringFromAscii(UInt8[] ascii);
```

Comment:
>UInt8[] [](start = 40, length = 7)

can we use string_view for these?

Reply:
No, the string_view is not a type allowed in IDL. Passing of arrays is implemented as array_view, which is pretty much the same as string_view. We are getting the same no-allocation benefits.

---
In reply to: [455235904](https://github.com/microsoft/react-native-windows/pull/5461#discussion_r455235904) [](ancestors = 455235904)